### ISSUE TITLE
open Website default page stops all applications

### ISSUE THREAD

**Issue by diliger** [StarcounterPrefabs/Website#24]
when I click the button Open for **Website** app in **Starcounter Administrator** 
page for `localhost.../Website` opens, results to `URI not found` and all started applications die.

is that appropriate behaviour? 
![image](https://cloud.githubusercontent.com/assets/15857369/22742475/e853530e-ee27-11e6-9517-2595e72f70af.png)

**Comment by warpech** [comment]
@diliger thanks for reporting! It is certainly not appropriate.

When reporting bugs, please provide information like:

- Starcounter version
- App version
- Error from the error log

---

I **can** reproduce it with Website app (`develop` branch) and Starcounter 2.3.0.5512

In Administrator Log I can see an exception:

```
Non database type Starcounter.Metadata.MetadataEntity of property Type in class Starcounter.Internal.Metadata.SetSpecifier
```

@andwah, @per-samuelsson any ideas what is happening?

**Comment by andwah** [comment]
Sorry, no ideas just by reading this issue.

**Comment by per-samuelsson** [comment]
> In Administrator Log I can see an exception:
>> Non database type Starcounter.Metadata.MetadataEntity of property Type in class Starcounter.Internal.Metadata.SetSpecifier

As an *exception*, really? Otherwise I would have suggested that thing was not related to the crash of the host, because it's already a [known issue](https://github.com/Starcounter/Starcounter/issues/3937). (But apparently by design)

**Comment by warpech** [comment]
@per-samuelsson you a right, this is an unrelated warning...

So yes, this is a crash that is caused by the handler `/Website` in the Website app. It happens also when you visit the URI directly, not from Administrator.

Visual Studio shows me that calling the handler throws:

```
System.StackOverflowException was unhandled
Message: An unhandled exception of type 'System.StackOverflowException' occurred in Starcounter.XSON.dll
```

The above exception is not registered in the error log.

After that, all running applications stop.

@diliger can you investigate what line causes the StackOverflowException?

**Comment by chrhol** [comment]
I will take a look as well. Sounds like some internal problem in Starcounter.XSON or that a sibling (=merged json) is attached in a way that causes an infinite loop.

I have a point to add checks for this inside TypedJSON to avoid stackoverflow, but haven't gotten around to implement it yet. I will raise the priority on it though.

**Comment by chrhol** [comment]
Something is messed up with the merged json/siblings

One instance of json in website is merged with a parent of itself.

```
x = {Website.LayoutPage}
y = {Website.ContainerPage}
(y.Parent.Parent.Parent.Parent.Parent == x) is true
```

I haven't found yet why this is happening, but this will lead to an infinite loop.

**Comment by diliger** [comment]
In debug mode this line causes to an error in
>	`Starcounter.XSON.dll!Starcounter.Json.GetTransactionHandle(bool lookInStepSiblings)`

![image](https://cloud.githubusercontent.com/assets/15857369/22774850/0778418a-eeba-11e6-95bf-ac70edbc61f7.png)

**Comment by per-samuelsson** [comment]
So that's why there was nothing in the log: stack overflows are generally hard to capture.

Now that @chrhol is on it, I feel calm. 👶

**Comment by chrhol** [comment]
From what I can see this happens because of the following: 

First on line 193 in `ContentHandlers.cs`:
```csharp
sectionJson.Rows[index].MergeJson(res.Resource as Json);
```
where 
`sectionJson.Rows[index]` is `{Website.ContainerPage}`
`res.Resource` is `{Website.LayoutPage}`

And then line 139 in the same class:
```csharp
master.TemplateModel = content;
```
where `master` is the same instance as the json that was merged on line 193 (`{Website.LayoutPage}`)

In this case, as @diliger pointed out the stackoverflow happens in `Starcounter.XSON.dll!Starcounter.Json.GetTransactionHandle(bool lookInStepSiblings)` but the overflow can come from other places. 

Basically: it should (cannot) be legal to connect two siblings where one is a parent of the other, or connect a json to parent that is a sibling further down in the tree.

The problem is that I know of no good (or bad for that matter) way to check this when they are connected. What I will try to do is to detect that we are stuck in a loop and try to print out a part of the jsontree and add that to the errormessage. That way there is a (small) chance at least we can see why and what happened.

**Comment by chrhol** [comment]
Added issue for adding detection: https://github.com/Starcounter/Starcounter/issues/4085

However, this will not solve anything. Just give a chance to make it easier to find out why it happened. The problems above still needs to be solved somehow. 

Why is master merged with a child if itself in this case?

**Comment by warpech** [comment]
I think that proper implementation of #29 will fix this

**Comment by warpech** [comment]
I will work on it this weekend.

**Comment by warpech** [comment]
I will unassign me from this because of other tasks that I should focus on. 

@diliger please give it another try after #30 is merged. It should be simple to fix this issue by renaming `/website/cms` to `/website`.

`/websiteprovider` should show some welcome text "Welcome to WebsiteProvider".

**Comment by diliger** [comment]
PR #38

**Comment by warpech** [comment]
Can you first review PR #35? I am afraid there might be some conflicts

**Comment by diliger** [comment]
I have only one question there 

> Shouldn't the launcher-menu be BEMified?

**Comment by diliger** [comment]
Merged. Closed.
